File: purecoincidence/src/main/scala/purecoincidence/Impurity.scala
package purecoincidence

object Impurity extends App {
  var current: Int = 0
  def impure(n: Int): Int = {
    current += n
    current
  }

  val i = impure(2)

  println((
    i + i,
    impure(2) + i,
    impure(2) + impure(2)
  ))  // (4,6,14)
}

File: hopefullypure/app/src/main/java/hopefullypure/Impurity.java
package hopefullypure;

import static java.lang.System.out;
import static java.lang.String.format;

public class Impurity {
    static int current = 0;
    static int impure(int n) {
        return current += n;
    }

    public static void main(String[] args) {
        int i = impure(2);
        out.println(
          format("%d, %d, %d",
                 i + i,
                 impure(2) + i,
                 impure(2) + impure(2)));
    }
}